### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

In [2]:
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [3]:
# Number of Purchases
df = pd.DataFrame(purchase_data)
pay_num = len(df[['Purchase ID']])
print(pay_num)

780


In [4]:
# Number of Players
# SN is the unique identifier for players. 
play_num = len(df['SN'].value_counts())
play_num

576

In [5]:
# Number of players as its own table
play_cnt = pd.DataFrame({'Number of Players':play_num}, index=[0])
play_cnt

Number of Players
0                576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [6]:
# Setting unique items, need to use Item ID rather than Item Name as some item names are duplicated, but have different item IDs
unqitem = len(df['Item ID'].value_counts())
# Setting the average price of all items sold
avgprice = df[['Price']].mean()
# Setting the total revenue of all items sold
totrev = df['Price'].sum()

# Setting a library to be input into a dataframe for the summary table, probably a redundant step
purch_analysis = {'Number of Purchases':play_num, 'Number of Unique Items':unqitem, 'Average Price':avgprice[0], 'Total Revenue':totrev}
purch_analysis

{'Number of Purchases': 576,
 'Number of Unique Items': 183,
 'Average Price': 3.050987179487176,
 'Total Revenue': 2379.77}

In [7]:
# Setting the first summary table
Sum_Table_1 = pd.DataFrame(purch_analysis, index=[0])

# Setting the format for the dollar values found in Summary Table 1
Sum_Table_1['Average Price'] = Sum_Table_1['Average Price'].map('${:,.2f}'.format)
Sum_Table_1['Total Revenue'] = Sum_Table_1['Total Revenue'].map('${:,.2f}'.format)
Sum_Table_1

Number of Purchases  Number of Unique Items Average Price Total Revenue
0                  576                     183         $3.05     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [8]:
# Defining the set of unique players by filtering by SN
gendf_nodup = df.drop_duplicates(['SN'])
gendf_nodup.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [9]:
# Defining gender count after removing the duplicate players 
gen_count = gendf_nodup['Gender'].value_counts()
gen_count

Male                     484
Female                    81
Other / Non-Disclosed     11
Name: Gender, dtype: int64

In [10]:
# Defining gender percentage
gen_per = gen_count/play_num*100

In [11]:
# Preparing the Second Summary Table
Sum_Table_2 = pd.DataFrame({'Total Count':gen_count, 'Percentage of Players':gen_per})
Sum_Table_2['Percentage of Players'] = Sum_Table_2['Percentage of Players'].map('{:,.2f}%'.format)
Sum_Table_2

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [12]:
gendf = df.groupby(['Gender'])

In [13]:
# Defining the sums spent by each gender
sumbygen = gendf['Price'].sum()
sumbygen

Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Price, dtype: float64

In [14]:
# The gender count with duplicates is still needed
gd_count = df['Gender'].value_counts()
gd_count

Male                     652
Female                   113
Other / Non-Disclosed     15
Name: Gender, dtype: int64

In [15]:
# To get the average price paid, we will need the sum of all prices divided by the counts of all genders, including the dupes
gen_avg = sumbygen / gd_count
gen_avg

Female                   3.203009
Male                     3.017853
Other / Non-Disclosed    3.346000
dtype: float64

In [16]:
# Defining the Average Total Purchase per Person
avgtot = sumbygen / gen_count
avgtot

Female                   4.468395
Male                     4.065372
Other / Non-Disclosed    4.562727
dtype: float64

In [17]:
# Preparing the Third Summary Table
Sum_Table_3 = pd.DataFrame({'Purchase Count':gd_count, 'Average Purchase Price':gen_avg, 'Total Purchase Value':sumbygen,
                           'Avg Total Purchase per Person':avgtot})
Sum_Table_3['Average Purchase Price'] = Sum_Table_3['Average Purchase Price'].map('${:,.2f}'.format)
Sum_Table_3['Total Purchase Value'] = Sum_Table_3['Total Purchase Value'].map('${:,.2f}'.format)
Sum_Table_3['Avg Total Purchase per Person'] = Sum_Table_3['Avg Total Purchase per Person'].map('${:,.2f}'.format)
Sum_Table_3

Purchase Count Average Purchase Price  \
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [18]:
# Establishing bins for ages
age_ranges = [0, 9, 14, 19, 24, 29, 34, 39, 100]

In [19]:
# Setting group names for the cut
group_names = ['<10','10-14','15-19','20-24','25-29','30-34','35-39','40+']

In [20]:
# Establishing age as bins
df['Age Range'] = pd.cut(df['Age'],age_ranges,labels=group_names)
df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Range  
0  Extraction, Quickblade Of Trembling Hands   3.53     20-24  
1                          Frenzied Scimitar   1.56       40+  
2                               Final Critic   4.88     20-24  
3                                Blindscythe   3.27     20-24  
4                                       Fury   1.44     20-24

In [21]:
# Rerunning the no duplicates
gendf_nodup = df.drop_duplicates(['SN'])

In [22]:
# Creating the dataframe grouped by age range with duplicates included
agedf = df.groupby(['Age Range'])

In [23]:
# Creating the dataframe grouped by age range with duplicates excluded
agedf_nodup = gendf_nodup.groupby(['Age Range'])

In [24]:
# Setting the age count with no duplicates
age_count = gendf_nodup['Age Range'].value_counts()

In [25]:
age_per = age_count / play_num * 100

In [26]:
# Preparing the Fourth Summary Table
Sum_Table_4 = pd.DataFrame({'Total Count':age_count, 'Percentage of Players':age_per})
Sum_Table_4['Percentage of Players'] = Sum_Table_4['Percentage of Players'].map('{:,.2f}%'.format)
Sum_Table_4.sort_index(ascending = [True, True, group_names])

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [27]:
# The age count with duplicates is the number of purchases by age
age_pur = df['Age Range'].value_counts()
age_pur

20-24    365
15-19    136
25-29    101
30-34     73
35-39     41
10-14     28
<10       23
40+       13
Name: Age Range, dtype: int64

In [28]:
# Defining the sums spent by each age range
sumbyage = agedf['Price'].sum()
sumbyage

Age Range
<10        77.13
10-14      82.78
15-19     412.89
20-24    1114.06
25-29     293.00
30-34     214.00
35-39     147.67
40+        38.24
Name: Price, dtype: float64

In [29]:
# To get the average price paid, we will need the sum of all prices divided by the counts of all age groups, including the dupes
age_avg_pr = sumbyage / age_pur
age_avg_pr 

10-14    2.956429
15-19    3.035956
20-24    3.052219
25-29    2.900990
30-34    2.931507
35-39    3.601707
40+      2.941538
<10      3.353478
dtype: float64

In [30]:
# Defining the Average Total Purchase per Person per age group
age_avgtot = sumbyage / age_count
age_avgtot

10-14    3.762727
15-19    3.858785
20-24    4.318062
25-29    3.805195
30-34    4.115385
35-39    4.763548
40+      3.186667
<10      4.537059
dtype: float64

In [31]:
Sum_Table_5 = pd.DataFrame({'Purchase Count':age_pur, 'Avg Purchase Price':age_avg_pr, 'Total Purchase Value':sumbyage, 
                            'Avg Total Purchase per Person':age_avgtot})
Sum_Table_5['Avg Purchase Price'] = Sum_Table_5['Avg Purchase Price'].map('${:,.2f}'.format)
Sum_Table_5['Total Purchase Value'] = Sum_Table_5['Total Purchase Value'].map('${:,.2f}'.format)
Sum_Table_5['Avg Total Purchase per Person'] = Sum_Table_5['Avg Total Purchase per Person'].map('${:,.2f}'.format)
Sum_Table_5.sort_index(ascending = [True, True, group_names])

Purchase Count Avg Purchase Price Total Purchase Value  \
<10                23              $3.35               $77.13   
10-14              28              $2.96               $82.78   
15-19             136              $3.04              $412.89   
20-24             365              $3.05            $1,114.06   
25-29             101              $2.90              $293.00   
30-34              73              $2.93              $214.00   
35-39              41              $3.60              $147.67   
40+                13              $2.94               $38.24   

      Avg Total Purchase per Person  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [32]:
# Setting a dataframe grouped by SN
sn_df = df.groupby(['SN'])

In [33]:
# Summing the whales' purchases
whale_tot = sn_df['Price'].sum()
whale_tot.head()

SN
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
Name: Price, dtype: float64

In [34]:
# Defining the number of purchases per whale
whale_count = sn_df['SN'].count()
whale_count.head()

SN
Adairialis76    1
Adastirin33     1
Aeda94          1
Aela59          1
Aelaria33       1
Name: SN, dtype: int64

In [35]:
# Defining the average purchase price per whale
whale_average = sn_df['Price'].mean()
whale_average.head()

SN
Adairialis76    2.28
Adastirin33     4.48
Aeda94          4.91
Aela59          4.32
Aelaria33       1.79
Name: Price, dtype: float64

In [36]:
# The sorting apparently keeps getting broken if dollar notation is applied before sorting
Sum_Table_6 = pd.DataFrame({'Purchase Count':whale_count,'Average Purchase Price':whale_average,'Total Purchase Value':whale_tot})
Sum_Table_6 = Sum_Table_6.sort_values(by='Total Purchase Value',ascending=False)

In [37]:
# Preparing the 6th Summary Table
Sum_Table_6['Average Purchase Price'] = Sum_Table_6['Average Purchase Price'].map('${:,.2f}'.format)
Sum_Table_6['Total Purchase Value'] = Sum_Table_6['Total Purchase Value'].map('${:,.2f}'.format)
Sum_Table_6.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [38]:
# Setting a dataframe defined by Item ID / Item Name
# There are a couple items with the same name
item_df = df.groupby(['Item ID', 'Item Name'])

In [39]:
# Summing the number of purchases per item
item_pur_count = item_df['Price'].count()
item_pur_count.head()

Item ID  Item Name         
0        Splinter              4
1        Crucifer              3
2        Verdict               6
3        Phantomlight          6
4        Bloodlord's Fetish    5
Name: Price, dtype: int64

In [40]:
# Taking the mean of the item price will just return the price of the item
item_price = item_df['Price'].mean()
item_price.head()

Item ID  Item Name         
0        Splinter              1.28
1        Crucifer              3.26
2        Verdict               2.48
3        Phantomlight          2.49
4        Bloodlord's Fetish    1.70
Name: Price, dtype: float64

In [41]:
# The total purchase value is just the sum of the purchases grouped by item
item_pur_total = item_df['Price'].sum()
item_pur_total.head()

Item ID  Item Name         
0        Splinter               5.12
1        Crucifer               9.78
2        Verdict               14.88
3        Phantomlight          14.94
4        Bloodlord's Fetish     8.50
Name: Price, dtype: float64

In [42]:
# Again we need to sort before applying the dollar notation in case the sort gets skewed
Sum_Table_7 = pd.DataFrame({'Purchase Count':item_pur_count,'Item Price':item_price,'Total Purchase Value':item_pur_total})
Sum_Table_7 = Sum_Table_7.sort_values(by='Purchase Count',ascending=False)

In [43]:
# Preparing the 7th summary table
Sum_Table_7['Item Price'] = Sum_Table_7['Item Price'].map('${:,.2f}'.format)
Sum_Table_7['Total Purchase Value'] = Sum_Table_7['Total Purchase Value'].map('${:,.2f}'.format)
Sum_Table_7.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [44]:
# This is just a variation of the previous table.
Sum_Table_8 = pd.DataFrame({'Purchase Count':item_pur_count,'Item Price':item_price,'Total Purchase Value':item_pur_total})
Sum_Table_8 = Sum_Table_8.sort_values(by='Total Purchase Value',ascending=False)


Sum_Table_8['Item Price'] = Sum_Table_8['Item Price'].map('${:,.2f}'.format)
Sum_Table_8['Total Purchase Value'] = Sum_Table_8['Total Purchase Value'].map('${:,.2f}'.format)
Sum_Table_8.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

# Report

## Observable trends

1. The playerbase for Heroes of Pymoli is overwhelmingly male at just over 84%. The data show that female players spend more on average than male players at \\$4.47 compared to males' \\$4.07. While appealing to female players may prove tricky with the remarkably tetchy demographic that is male gamers, it would likely be profitable. The developers should seek to maximize the profits from that portion of their market. 

2. With a max total purchase value by a single player of \\$18.96, Heroes of Pymoli could be doing a lot more to leverage the pocketbooks of their "whale" playerbase. A whale is the name applied to players who spend thousands of dollars on in-game content much like the items analyzed here. To make Heroes of Pymoli more profitable, the developers should probably provide more incentive for whales to spend their money by making the gear they purchase less durable or obsolete as players progress. While players may not like it, it has proved profitable for other developers of games with micro-transactions. 

3. The biggest chunk of money per age demographic is \\$1,114.06 from the 20-24 demographic with a fairly high \\$4.32 average amount spent per player. While 20-24 year-olds are a reliable demographic, more outreach should be done to entice older players with more disposable income to play and purchase items. 